metadata-ingestion/src/datahub/ingestion/source/qlik_sense/qlik_sense.py

```diff
@@ -0,0 +1,529 @@
+        dataset_urn = self._gen_qlik_dataset_urn(
+            self._get_app_dataset_identifier(dataset)
+        )
+
+        dataset_properties = DatasetProperties(
+            name=dataset.tableName,
+            qualifiedName=dataset.tableAlias,
+        )
+        dataset_properties.customProperties.update(
+            {
+                Constant.DATACONNECTORID: dataset.dataconnectorid,
+                Constant.DATACONNECTORNAME: dataset.dataconnectorName,
+            }
+        )
+
+        return MetadataChangeProposalWrapper(
+            entityUrn=dataset_urn, aspect=dataset_properties
+        ).as_workunit()
+
+    def _get_app_dataset_identifier(self, dataset: QlikAppDataset) -> str:
+        return f"{dataset.databaseName}.{dataset.schemaName}.{dataset.tableName}"
+
+    def _gen_app_dataset_workunit(
+        self, datasets: List[QlikAppDataset], app_id: str
+    ) -> Iterable[MetadataWorkUnit]:
+        for dataset in datasets:
+            dataset_identifier = self._get_app_dataset_identifier(dataset)
+            dataset_urn = self._gen_qlik_dataset_urn(dataset_identifier)
+
+            yield self._gen_entity_status_aspect(dataset_urn)
+
+            yield self._gen_schema_metadata(dataset_identifier, dataset.datasetSchema)
+
+            yield self._gen_app_dataset_properties(dataset)
+
+            yield from add_entity_to_container(
+                container_key=self._gen_app_key(app_id),
+                entity_type="dataset",
+                entity_urn=dataset_urn,
+            )
+
+            dpi_aspect = self._gen_dataplatform_instance_aspect(dataset_urn)
+            if dpi_aspect:
+                yield dpi_aspect
+
+            yield MetadataChangeProposalWrapper(
+                entityUrn=dataset_urn,
+                aspect=SubTypes(typeNames=[DatasetSubTypes.QLIK_DATASET]),
+            ).as_workunit()
+
+            yield self._gen_app_dataset_upstream_lineage(dataset)
+
+    def _gen_app_workunit(self, app: App) -> Iterable[MetadataWorkUnit]:
+        """
+        Map Qlik App to Datahub container
+        """
+        yield from gen_containers(
+            container_key=self._gen_app_key(app.id),
+            name=app.qTitle,
+            description=app.description,
+            sub_types=[BIContainerSubTypes.QLIK_APP],
+            parent_container_key=self._gen_space_key(app.spaceId),
+            extra_properties={Constant.QRI: app.qri, Constant.USAGE: app.qUsage},
+            owner_urn=builder.make_user_urn(app.ownerId)
+            if self.config.ingest_owner
+            else None,
+            created=int(app.createdAt.timestamp() * 1000),
+            last_modified=int(app.updatedAt.timestamp() * 1000),
+        )
+        yield from self._gen_sheets_workunit(app.sheets, app.spaceId, app.id)
+        yield from self._gen_app_dataset_workunit(app.datasets, app.id)
+
+    def _gen_qlik_dataset_urn(self, dataset_identifier: str) -> str:
```

Comment:
I'm not sure this naming schema makes sense, since urns aren't guaranteed to be unique. e.g. if the same snowflake table is used in two spaces, those will conflict

qlik generates `qri`s for every qlik dataset right? is there a reason we're not using those?

Reply:
if the same snowflake table is used in two spaces, then I guess we should have same dataset entity present in two spaces containers.
By the way Is it possible?

Reply:
> if the same snowflake table is used in two spaces, then I guess we should have same dataset entity present in two spaces containers.

Because Qlik allows you to do some transformations and things when creating the Qlik table from the Snowflake table, we actually _don't_ want this. Instead, we want one Qlik dataset per space, with all having upstreamLineage pointing at the snowflake table.